File: frontend/scripts/audit-offline.mjs
/**
 * Fail the build/test if frontend runtime files pull remote internet resources.
 * localhost / 127.0.0.1 are allowed.
 *
 * Dist bundles may contain vendor license/comment URLs from lightweight-charts.
 * Those strings are not fetched at runtime. The TradingView attribution logo is
 * disabled in application source (layout.attributionLogo = false).
 */
import { readdirSync, readFileSync, statSync } from "node:fs";
import { extname, join, relative } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = fileURLToPath(new URL("..", import.meta.url));
const SOURCE_TARGETS = ["src", "index.html", "public"];
const DIST_TARGETS = ["dist"];
const TEXT_EXT = new Set([
  ".html",
  ".js",
  ".jsx",
  ".ts",
  ".tsx",
  ".css",
  ".json",
  ".svg",
  ".mjs",
  ".cjs",
]);

const FORBIDDEN = [
  /fonts\.googleapis/i,
  /fonts\.gstatic/i,
  /api\.binance/i,
  /data\.binance/i,
  /stream\.binance/i,
  /cdn\.jsdelivr/i,
  /unpkg\.com/i,
  /cdnjs\.cloudflare/i,
];

const HTTP_URL = /https?:\/\/[^\s"'`>]+/gi;
const PROTOCOL_RELATIVE = /(?:['"`(=])\/\/(?!\/)[^\s"'`>]+/g;
const LOCAL = /^https?:\/\/(?:127\.0\.0\.1|localhost)(?::\d+)?(?:\/|$)/i;
const XMLNS = /^https?:\/\/www\.w3\.org\/\d{4}\/(?:svg|xlink)$/i;
const VENDOR_DIST_URL = /^https?:\/\/(?:www\.)?(?:apache\.org|w3\.org|fileformat\.info|bugs\.chromium\.org|developer\.apple\.com|www\.tradingview\.com|reactjs\.org|react\.dev)\b/i;

function walk(start) {
  const files = [];
  const stack = [start];
  while (stack.length > 0) {
    const current = stack.pop();
    if (!current) {
      continue;
    }
    let stats;
    try {
      stats = statSync(current);
    } catch {
      continue;
    }
    if (stats.isDirectory()) {
      for (const entry of readdirSync(current)) {
        if (entry === "node_modules") {
          continue;
        }
        stack.push(join(current, entry));
      }
    } else if (stats.isFile()) {
      files.push(current);
    }
  }
  return files;
}

function collect(targets) {
  const files = [];
  for (const target of targets) {
    const abs = join(ROOT, target);
    try {
      const stats = statSync(abs);
      if (stats.isDirectory()) {
        files.push(...walk(abs));
      } else if (stats.isFile()) {
        files.push(abs);
      }
    } catch {
      // Optional paths such as dist/ may not exist yet.
    }
  }
  return files;
}

function scan(files, { allowVendorUrls }) {
  const violations = [];
  for (const file of files) {
    if (!TEXT_EXT.has(extname(file).toLowerCase()) && !file.endsWith("index.html")) {
      continue;
    }
    const text = readFileSync(file, "utf8");
    const rel = relative(ROOT, file);
    for (const pattern of FORBIDDEN) {
      if (pattern.test(text)) {
        violations.push(`${rel}: forbidden token ${pattern}`);
      }
    }
    if (!allowVendorUrls && /tradingview/i.test(text)) {
      violations.push(`${rel}: tradingview token in application source`);
    }
    for (const match of text.matchAll(HTTP_URL)) {
      const url = match[0];
      if (LOCAL.test(url) || XMLNS.test(url)) {
        continue;
      }
      if (allowVendorUrls && VENDOR_DIST_URL.test(url)) {
        continue;
      }
      violations.push(`${rel}: remote URL ${url}`);
    }
    for (const match of text.matchAll(PROTOCOL_RELATIVE)) {
      violations.push(`${rel}: protocol-relative URL ${match[0]}`);
    }
  }
  return violations;
}

const sourceFiles = collect(SOURCE_TARGETS);
if (sourceFiles.length === 0) {
  console.error("Offline audit failed: no frontend source files found.");
  process.exit(1);
}

const sourceText = sourceFiles
  .filter((file) => file.endsWith("TradingChart.tsx"))
  .map((file) => readFileSync(file, "utf8"))
  .join("\n");
if (!sourceText.includes("attributionLogo: false")) {
  console.error("Offline audit failed: chart source must disable layout.attributionLogo.");
  process.exit(1);
}

const violations = [
  ...scan(sourceFiles, { allowVendorUrls: false }),
  ...scan(collect(DIST_TARGETS), { allowVendorUrls: true }),
];

if (violations.length > 0) {
  console.error("Offline audit failed:");
  for (const line of violations) {
    console.error(`  ${line}`);
  }
  process.exit(1);
}

console.log(`Offline audit passed (${sourceFiles.length} source files scanned).`);

File: scripts/audit_offline.py
"""Scan frontend runtime files for remote internet dependencies."""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

FORBIDDEN = [
    re.compile(r"fonts\.googleapis", re.I),
    re.compile(r"fonts\.gstatic", re.I),
    re.compile(r"api\.binance", re.I),
    re.compile(r"data\.binance", re.I),
    re.compile(r"stream\.binance", re.I),
    re.compile(r"cdn\.jsdelivr", re.I),
    re.compile(r"unpkg\.com", re.I),
    re.compile(r"cdnjs\.cloudflare", re.I),
]
HTTP_URL = re.compile(r"https?://[^\s\"'`>]+", re.I)
PROTOCOL_RELATIVE = re.compile(r"[\'\"`(=]//(?!/)[^\s\"'`>]+")
LOCAL_HTTP = re.compile(r"^https?://(?:127\.0\.0\.1|localhost)(?::\d+)?(?:/|$)", re.I)
XMLNS = re.compile(r"^https?://www\.w3\.org/\d{4}/(?:svg|xlink)$", re.I)
VENDOR_DIST_URL = re.compile(
    r"^https?://(?:www\.)?(?:apache\.org|w3\.org|fileformat\.info|"
    r"bugs\.chromium\.org|developer\.apple\.com|www\.tradingview\.com|"
    r"reactjs\.org|react\.dev)\b",
    re.I,
)
TEXT_SUFFIXES = {".html", ".js", ".jsx", ".ts", ".tsx", ".css", ".json", ".svg", ".mjs", ".cjs"}


def iter_files(frontend: Path, relatives: tuple[str, ...]) -> list[Path]:
    files: list[Path] = []
    for relative in relatives:
        path = frontend / relative
        if path.is_file():
            files.append(path)
        elif path.is_dir():
            files.extend(child for child in path.rglob("*") if child.is_file())
    return files


def scan(frontend: Path, files: list[Path], *, allow_vendor_urls: bool) -> list[str]:
    violations: list[str] = []
    for path in files:
        if path.suffix.lower() not in TEXT_SUFFIXES and path.name != "index.html":
            continue
        try:
            text = path.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            continue
        rel = path.relative_to(frontend.parent)
        for pattern in FORBIDDEN:
            if pattern.search(text):
                violations.append(f"{rel}: forbidden token {pattern.pattern}")
        if not allow_vendor_urls and re.search(r"tradingview", text, re.I):
            violations.append(f"{rel}: tradingview token in application source")
        for match in HTTP_URL.finditer(text):
            url = match.group(0)
            if LOCAL_HTTP.match(url) or XMLNS.match(url):
                continue
            if allow_vendor_urls and VENDOR_DIST_URL.match(url):
                continue
            violations.append(f"{rel}: remote URL {url}")
        for match in PROTOCOL_RELATIVE.finditer(text):
            violations.append(f"{rel}: protocol-relative URL {match.group(0)}")
    return violations


def audit(frontend: Path) -> list[str]:
    source_files = iter_files(frontend, ("src", "index.html", "public"))
    dist_files = iter_files(frontend, ("dist",))
    chart_source = frontend / "src" / "components" / "chart" / "TradingChart.tsx"
    violations: list[str] = []
    if not chart_source.exists() or "attributionLogo: false" not in chart_source.read_text(encoding="utf-8"):
        violations.append("chart source must disable layout.attributionLogo")
    violations.extend(scan(frontend, source_files, allow_vendor_urls=False))
    violations.extend(scan(frontend, dist_files, allow_vendor_urls=True))
    return violations


def main() -> int:
    parser = argparse.ArgumentParser(description="Offline runtime asset audit")
    parser.add_argument(
        "--frontend",
        type=Path,
        default=Path(__file__).resolve().parents[1] / "frontend",
    )
    args = parser.parse_args()
    if not args.frontend.exists():
        print(f"frontend not found: {args.frontend}", file=sys.stderr)
        return 2
    violations = audit(args.frontend)
    if violations:
        print("Offline audit failed:")
        for line in violations:
            print(f"  {line}")
        return 1
    print("Offline audit passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
